### PR TITLE
Various optimizations of the network code

### PR DESCRIPTION
**Pull request by stevenroose**
It would be really useful to have the `command` method on NetworkMessage and it's strange it wasn't like that in the first place because it seems a RawNetworkMessage would always be unwrapped into a NetworkMessage anyway.

Also the CommandString is 99.99% of the time gonna be a `&'static str`, so a bit silly to have a `String` for that. I made it into a `Cow` instead.

@tamasblummer I believe you did a lot of the wire code in rust-bitcoin, could you review?

### CODE REVIEW

**Comment by elichai** [comment]
ConceptACK. are people supposed to somehow access the string inside of `CommandString`?
because you made the field inside of it private when before it was public (not saying it should be public, just asking)

**Comment by stevenroose** [comment]
There is `.to_string()` for that. But it wouldn't be useful in any situation I can think of. I can't see a situation where you want to take ownership over the inside (a la into_inner).

In fact one addition I could think of is to change the `From<String>` into `TryFrom` because there is a max length of 8 characters IIRC. But `TryFrom` is only 1.34-stable. Not yet, I guess.

**Comment by stevenroose** [comment]
Should I generalize the `Encodable` and `Decodable` implementations of `String` and `Cow<'static, str>` into one implementation for `T: AsRef<str>`?

**Comment by stevenroose** [comment]
Ok no I shouldn't.

**Comment by stevenroose** [comment]
I added a small commit to add `Copy` to the `InvType` enum. I also rebased.

**Review by apoelstra — DISMISSED** [review]
utack. Would definitely like Tamas' ack before merging

File: src/network/message.rs [hunk 1]
```diff
@@ -34,17 +35,40 @@ use consensus::encode::MAX_VEC_SIZE;
...
     fn consensus_encode<S: io::Write>(
         &self,
         s: S,
     ) -> Result<usize, encode::Error> {
-        let &CommandString(ref inner_str) = self;
         let mut rawbytes = [0u8; 12];
-        let strbytes = inner_str.as_bytes();
+        let strbytes = self.0.as_bytes();
         if strbytes.len() > 12 {
             panic!("Command string longer than 12 bytes");
```
Comment (tamasblummer) [inline src/network/message.rs]: Would it not be an easy attack vector to let the node panic?

**Review by tamasblummer — CHANGES_REQUESTED** [review]

File: src/network/message.rs (hunk 1 above)
Comment (stevenroose) [reply src/network/message.rs]: That's a good point. TryFrom is Rust 1.34. Hmm yeah then I propose we just truncate them. We already have this code that uses the 12-byte array. I'll add a commit.

Comment (stevenroose) [reply src/network/message.rs]: I added another commit with a test serializing an oversized command string. I think is best like this. Lateron when we have a MSRV of > 1.34, we can upgrade to TryFrom.

**Comment by apoelstra** [comment]
I think we should return an error rather than truncating. Truncation creates a malleability vector and will prevent round-tripping network messages in fuzz tests.

**Comment by stevenroose** [comment]
Fixed.

Btw, another case for `PartialEq` on `Error`s :)
```rust
        assert_eq!(
            CommandString("AndrewAndrewAndrew".into()).consensus_encode(&mut encoder).err(),
            Some(encode::Error::UnrecognizedNetworkCommand("AndrewAndrewAndrew".into()))
        );
```

**Review by apoelstra — APPROVED** [review]
new changes lgtm

**Review by tamasblummer — APPROVED** [review]

**Comment by apoelstra** [comment]
Blocked codecov-commenter from the org.

**Pull request merged**